File: src/test/scala/com/exasol/cloudetl/kafka/deserialization/DeserializationFactoryTest.scala
package com.exasol.cloudetl.kafka.deserialization

import com.exasol.cloudetl.kafka.KafkaConsumerProperties

import org.scalatest.funsuite.AnyFunSuite

class DeserializationFactoryTest extends AnyFunSuite {

  test("must provide default serializers: string key and avro value, fields not specified") {
    val properties = new KafkaConsumerProperties(Map("SCHEMA_REGISTRY_URL" -> "someUrl"))
    val deserializers =
      DeserializationFactory.getSerializers(Seq(RecordKey, RecordValueFields), properties)

    assert(deserializers.keyDeserializer.isInstanceOf[AsStringDeserializer])
    assert(deserializers.valueDeserializer.isInstanceOf[GenericRecordDeserializer])
  }

  test("must not deserialize key when not specified in fieldSpecs") {
    val properties = new KafkaConsumerProperties(Map("RECORD_FORMAT" -> "json"))
    val deserializers = DeserializationFactory.getSerializers(Seq(RecordValue), properties)

    assert(deserializers.keyDeserializer === IgnoreKeyDeserializer)
    assert(deserializers.valueDeserializer.isInstanceOf[JsonDeserializer])
  }

  test("key must be ignored when ot requested") {
    val params = Map("RECORD_KEY_FORMAT" -> "avro", "RECORD_VALUE_FORMAT" -> "string")
    val properties = new KafkaConsumerProperties(params)
    val deserializers = DeserializationFactory.getSerializers(Seq(RecordValue), properties)

    assert(deserializers.keyDeserializer === IgnoreKeyDeserializer)
    assert(deserializers.valueDeserializer.isInstanceOf[AsStringDeserializer])
  }

  test("key must be taken into account when ot requested") {
    val params = Map("RECORD_KEY_FORMAT" -> "json", "RECORD_VALUE_FORMAT" -> "string")
    val properties = new KafkaConsumerProperties(params)
    val deserializers = DeserializationFactory.getSerializers(
      Seq(RecordKeyField("someField"), RecordValue),
      properties
    )

    assert(deserializers.keyDeserializer.isInstanceOf[JsonDeserializer])
    assert(deserializers.valueDeserializer.isInstanceOf[AsStringDeserializer])
  }
}

File: src/test/java/com/exasol/cloudetl/kafka/deserialization/DeserializationFactoryTest.java
package com.exasol.cloudetl.kafka.deserialization;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.containsString;
import static org.junit.jupiter.api.Assertions.*;

import java.util.List;
import java.util.Map;

import org.apache.kafka.common.serialization.Deserializer;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.junit.jupiter.api.Test;

import com.exasol.cloudetl.kafka.KafkaConnectorException;
import com.exasol.cloudetl.kafka.KafkaConsumerProperties;

import nl.jqno.equalsverifier.EqualsVerifier;
class DeserializationFactoryTest {
    @Test
    void providesDefaultSerializersStringKeyAndAvroValueWhenFieldsAreNotSpecified() {
        final var properties = new KafkaConsumerProperties(Map.of("SCHEMA_REGISTRY_URL", "someUrl"));
        final var deserializers = DeserializationFactory
                .getSerializers(List.of(RecordKey.INSTANCE, RecordValueFields.INSTANCE), properties);

        assertInstanceOf(AsStringDeserializer.class, deserializers.keyDeserializer);
        assertInstanceOf(GenericRecordDeserializer.class, deserializers.valueDeserializer);
    }

    @Test
    void doesNotDeserializeKeyWhenNotSpecifiedInFieldSpecs() {
        final var properties = new KafkaConsumerProperties(Map.of("RECORD_FORMAT", "json"));
        final var deserializers = DeserializationFactory.getSerializers(List.of(RecordValue.INSTANCE), properties);

        assertSame(IgnoreKeyDeserializer.INSTANCE, deserializers.keyDeserializer);
        assertInstanceOf(JsonDeserializer.class, deserializers.valueDeserializer);
    }

    @Test
    void ignoresKeyWhenNotRequested() {
        final var properties = new KafkaConsumerProperties(
                Map.of("RECORD_KEY_FORMAT", "avro", "RECORD_VALUE_FORMAT", "string"));
        final var deserializers = DeserializationFactory.getSerializers(List.of(RecordValue.INSTANCE), properties);

        assertSame(IgnoreKeyDeserializer.INSTANCE, deserializers.keyDeserializer);
        assertInstanceOf(AsStringDeserializer.class, deserializers.valueDeserializer);
    }

    @Test
    void takesKeyIntoAccountWhenRequested() {
        final var properties = new KafkaConsumerProperties(
                Map.of("RECORD_KEY_FORMAT", "json", "RECORD_VALUE_FORMAT", "string"));
        final var deserializers = DeserializationFactory
                .getSerializers(List.of(new RecordKeyField("someField"), RecordValue.INSTANCE), properties);

        assertInstanceOf(JsonDeserializer.class, deserializers.keyDeserializer);
        assertInstanceOf(AsStringDeserializer.class, deserializers.valueDeserializer);
    }

    @Test
    void failsForUnsupportedRecordFormats() {
        final KafkaConnectorException thrown = assertThrows(KafkaConnectorException.class,
                () -> DeserializationFactory.getDeserialization("protobuf"));

        assertAll(() -> assertThat(thrown.getMessage(), containsString("E-KCE-19")),
                () -> assertThat(thrown.getMessage(), containsString("not supported")));
    }

    @Test
    void recordDeserializersEqualsAndHashCodeFollowContract() {
        final Deserializer<Map<FieldSpecification, List<Object>>> red = IgnoreKeyDeserializer.INSTANCE;
        final Deserializer<Map<FieldSpecification, List<Object>>> black = new JsonDeserializer(
                List.of(new RecordValueField("field")), new StringDeserializer());

        EqualsVerifier.forClass(DeserializationFactory.RecordDeserializers.class)
                .withPrefabValues(Deserializer.class, red, black)
                .verify();
    }
}
